### DISCUSSION TITLE
linearization vs. optimization

### DISCUSSION THREAD

**Discussion by andrisi** [qpdf/qpdf#919] [Ideas]
Perhaps its relevant to QPDF: https://github.com/pdfcpu/pdfcpu/discussions/578 - I'm using QPDF to linearize files, but it seems to reverse the optimization done by other tools. Is there a way to combine the benefits of the two at least partially? So there is a 10mb file, which is reduced the 600K by optimization, but becomes 10mb again with linearization.

**Comment by jberkenbilt** [comment]
I don't see how qpdf could make a 600K file go back to 10mb again. I'd have to see an example to understand what's going on. Other than the slight increase in size from adding hint tables and forcing a few additional object types to not be compressed, linearization should have a minimal impact on file size. I agree with the commenter in the other ticket though: linearization isn't really worth it these days. In fact, at one time, Google used to send files through qpdf to linearize them when exporting from Google Docs, but they no longer do this. There are still use cases for linearizing large files that are expected to be accessed behind slow connections. qpdf's linearization was originally written exactly for that purpose (20 years ago in support of a Ford Foundation project in sub-Saharan Africa).

**Comment by jberkenbilt** [comment]
I could probably invent an example of where qpdf would do this, but it would be kind of pathological. Page dictionaries and the document catalog are not allowed to be compressed in linearized PDFs, so if there were a page dictionary or document catalog that contained a very large amount of data, this could cause this, but that would be pretty unusual. If it turns out to be something like that, my advice would be to not linearize if linearization increases the file size beyond a certain amount. There is definitely no value to linearizing a 600K file though. To construct a file that would exhibit this behavior, I would add an array to the document catalog or page dictionaries that had a particular value repeated a few million times. This would compress down to almost nothing, but if uncompressed, it would be very large. I have actually seen files that have arrays of that type, but not generally in page dictionaries or the document catalog.

**Comment by andrisi** [comment]
Sure, a 600K file is just fine as it is. I'll collect a batch of examples. I have docs that are 100-300mb, and optimization halves their size, so its certainly very useful, and I'll test how it plays together with linearization, as thats still useful at these doc sizes. Thanks for the detailed answers, and I'll get back with specific examples a bit later.

**Reply by jberkenbilt** [reply]
Great. If we find some pattern, I'm not averse to tweaking qpdf's linearization if there's a way to do it, such as pulling certain parts out of the page dictionary and making them indirect, but let's see what you come up with first.

**Reply by XXWHAhZo** [reply]
Encountered a 2MB file observing just a small increase of 5KB by linearization. Want me to share it to you?
